javalib/src/main/scala/java/util/NaturalComparator.scala

```diff
@@ -12,27 +12,35 @@
 
 package java.util
 
-/** A universal `Comparator` using the *natural ordering* of the elements.
+/** Utility for *natural ordering* of elements.
  *
  *  A number of JDK APIs accept a possibly-null `Comparator`, and use the
  *  *natural ordering* of elements when it is `null`. This universal comparator
  *  can be used internally instead of systematically testing for `null`,
  *  simplifying code paths.
  *
- *  The `compare()` method of this comparator throws `ClassCastException`s when
- *  used with values that cannot be compared using natural ordering, assuming
- *  Scala.js is configured with compliant `asInstanceOf`s. The behavior is
- *  otherwise undefined.
+ *  The methods use the fact that Comparator.naturalOrder actually always
+ *  returns the same object. That `compare()` method of this comparator throws
+ *  `ClassCastException`s when used with values that cannot be compared using
+ *  natural ordering, assuming Scala.js is configured with compliant
+ *  `asInstanceOf`s. The behavior is otherwise undefined.
  */
-private[util] object NaturalComparator extends Comparator[Any] {
-  def compare(o1: Any, o2: Any): Int =
-    o1.asInstanceOf[Comparable[Any]].compareTo(o2)
+private[util] object NaturalComparator {
+  /** Returns a natural comparator even for types that are not statically
+   *  known to be <: Comparable.
+   */
+  @inline
+  def forAnyType[A](): Comparator[A] = {
+    // Any type argument that is comparable will do.
+    Comparator.naturalOrder[String]().asInstanceOf[Comparator[A]]
+  }
```

Comment:
I don't like that we're using `String` here. Consider adding the following in `Comparator` to make things less ugly:
```scala
  @inline
  private[util] def naturalOrderForAnyType[A](): Comparator[A] =
    ReusableNaturalOrder.asInstanceOf[Comparator[A]]
```
and call that from this `forAnyType` method. This way, we don't have to go through a specific non-`Any` type, and the viability of the casts stays closer to the original definition.

In fact, even `def naturalOrder()` itself can delegate to `naturalOrderForAnyType`, in that case.

Reply:
Yes, fair enough. I have opted to just use `NaturalComparator` as it is. That also gets rid of the strange contract that `forAnyType` must return the same object all the time (for `select` / `unselect`).